### DISCUSSION TITLE
Resize the canvas content based on the device screen / container width

### DISCUSSION THREAD

**Discussion by Angelo-Sterlo** [vinothpandian/react-sketch-canvas#116] [Q&A]
If I draw something on the canvas from a desktop device, when I upload the path on a mobile react-sketch-canvas, it cuts out a part of the drawing, showing only the top left corner of the content.

Canvas drawing from desktop:
![image](https://user-images.githubusercontent.com/60478550/218498998-97570dcb-5402-4a77-985f-5da44c9158d1.png)

Canvas Drawing from Mobile:
![image](https://user-images.githubusercontent.com/60478550/218499409-fe4be556-4734-45a6-8c0a-6ef63ab91d8f.png)

I'm currently working with Ionic-React/Typescript.

Is there a way to keep the content visible when changing the canvas width and height?

Thank you in advance!

**Comment by vinothpandian** [comment]
Hey @Angelo-Sterlo,

You can use the [latest beta version](https://www.npmjs.com/package/react-sketch-canvas/v/7.0.0-next.2) to achieve this. 

You'll have to pass `withViewBox` property to `ReactSketchCanvas` to make it resizable.

**Reply by Angelo-Sterlo** [reply]
I tried but I still have the same problem, if I draw something on the web version, it's cut off in the mobile version.
The same problem happens if I first draw on mobile, the web version shows it in the top left corner of the canvas.

Is there another solution? Any CSS class that I can use?

**Reply by vinothpandian** [reply]
That's strange. Maybe your issue might be related to #114. I'll push a quick fix for that soon.

As far as I know, the only solution is to set the [viewBox](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/viewBox) property.

**Reply by dereklance** [reply]
@vinothpandian  for me, the `withViewBox` property works to scale everything up/down on resize. but then when you add another stroke to the canvas after resizing, all strokes on the canvas "shift" and I encounter the same problem where the strokes all get cut off

**Reply by vinothpandian** [reply]
@Angelo-Sterlo @dereklance 

Added a new issue to track this #117

**Reply by Angelo-Sterlo** [reply]
I've managed to make the screen orientation landscape, this helps to keep the aspect ratio of the canvas and the content fits perfectly, the problem is that I want to scale the entire view down, in order for the user to be able to draw on the canvas without scrolling up and down.
This causes the cursor on the canvas to shift when drawing as @dereklance said.

**Comment by GeoTimber** [comment]
Can anyone give any hints how this could possibly be fixed or if it is going to be really hard to fix?

I dont mind having a go but a little bit of background info could be usefull.

withViewBox works until you start to draw as described above

thank you anyway! 👏👏👏👏

**Reply by GeoTimber** [reply]
Looks like the paths in the stack need re-scaling on resize? Are they drawn again from the stack with the wrong scale?

**Reply by GeoTimber** [reply]
Ok I think the easiest solution would be add an optional viewboxWidth and viewboxHeight prop for the svg if you want to use withViewBox

Then you can use those to specify your canvas resolution, and set it once when you know your aspect ratio when loading the component.

The getCoordinates function would need a little rewrite to use the correct coordinates adjusted to the viewbox. this seems to work in my setup
```js 
 const getCoordinates = 
    (pointerEvent: React.PointerEvent<HTMLDivElement>): Point => {
      const boundingArea = canvasRef.current?.getBoundingClientRect();
      canvasSizeRef.current = boundingArea
        ? {
            width: boundingArea.width,
            height: boundingArea.height,
          }
        : null;

      const scrollLeft = window.scrollX ?? 0;
      const scrollTop = window.scrollY ?? 0;

      if (!boundingArea) {
        return { x: 0, y: 0 };
      }

      const coordinates =  {
        x: pointerEvent.pageX - boundingArea.left - scrollLeft,
        y: pointerEvent.pageY - boundingArea.top - scrollTop,
      };

      if (props.withViewBox && props.viewBoxHeight && props.viewBoxWidth) {
        const scaleX = boundingArea.width / props.viewBoxWidth;
        const scaleY = boundingArea.height / props.viewBoxHeight;
         // Height scaling factor
        const viewBoxCoordinates = {
          x: coordinates.x / scaleX,
          y: coordinates.y / scaleY,
        };
        return viewBoxCoordinates;
      } else {
        return coordinates;
      }
    }
```

And the viewbox would need to use these as well

```js
        viewBox={
          // eslint-disable-next-line no-nested-ternary
          withViewBox
          ? (viewBoxWidth && viewBoxHeight)
            ? `0 0 ${viewBoxWidth} ${viewBoxHeight}`
            : (canvasSizeRef.current !== null)
              ? `0 0 ${canvasSizeRef.current.width} ${canvasSizeRef.current.height}`
              : undefined
          : undefined
```

**Reply by xjamundx** [reply]
You got a fork with those anywhere?

**Reply by GeoTimber** [reply]
> You got a fork with those anywhere?


Unfortunately my setup didnt allow me to fork it easily and I am doing some other things, I also vaguely remember having to change the import paths a little bit to make it fit into my webpack cordova project with all the changes needed.

But here you have the code that I am using and  works for me and allows rescale
[reactSketchCanvas.tar.gz](https://github.com/user-attachments/files/17477093/reactSketchCanvas.tar.gz)

**Reply by xjamundx** [reply]
Thanks for sharing, not working for me out of the box, but I'll keep poking at it, thanks! Maybe I've got the wrong values for viewBoxWidth/Height 🤔 

https://github.com/vinothpandian/react-sketch-canvas/pull/178/files

**Reply by GeoTimber** [reply]
You can see it working here, click on the clapping hands to take a screenshot
https://checkin.dev.poqueiraweb.com/

and here is a video

https://github.com/user-attachments/assets/fc0a2148-1c57-4bd2-b7c7-a281ecdfc75a

**Reply by xjamundx** [reply]
Got it! Thanks for sharing the video. I just had already applied the zoom to the viewBox width/height instead of leaving it as the original size
